src/codegen/encoding/scale_encoding.rs

```diff
@@ -1,66 +1,196 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::codegen::cfg::{ControlFlowGraph, Instr};
-use crate::codegen::encoding::AbiEncoding;
+use crate::codegen::encoding::{increment_by, AbiEncoding};
 use crate::codegen::vartable::Vartable;
 use crate::codegen::{Builtin, Expression};
-use crate::sema::ast::{Namespace, Parameter, RetrieveType, Type};
+use crate::sema::ast::{Namespace, Parameter, Type, U32};
 use solang_parser::pt::Loc;
+use std::collections::HashMap;
+
+use super::LOC;
 
-/// This struct implements the trait AbiEncoding for Parity's Scale encoding
 pub(super) struct ScaleEncoding {
-    /// Are we pakced encoding?
+    storage_cache: HashMap<usize, Expression>,
     packed_encoder: bool,
 }
 
 impl ScaleEncoding {
-    pub fn new(packed: bool) -> ScaleEncoding {
-        ScaleEncoding {
+    pub fn new(packed: bool) -> Self {
+        Self {
+            storage_cache: HashMap::new(),
             packed_encoder: packed,
         }
     }
 }
 
-impl AbiEncoding for ScaleEncoding {
-    fn abi_encode(
-        &mut self,
-        loc: &Loc,
-        mut args: Vec<Expression>,
-        _ns: &Namespace,
-        vartab: &mut Vartable,
-        cfg: &mut ControlFlowGraph,
-    ) -> (Expression, Expression) {
-        let tys = args.iter().map(|e| e.ty()).collect::<Vec<Type>>();
+/// Encode `expr` into `buffer` as a compact integer. More information can found in the
+/// [SCALE documentation](https://docs.substrate.io/reference/scale-codec/).
+fn encode_compact(
+    expr: &Expression,
+    buffer: Option<&Expression>,
+    offset: Option<&Expression>,
+    vartab: &mut Vartable,
+    cfg: &mut ControlFlowGraph,
+) -> Expression {
+    let small = cfg.new_basic_block("small".into());
```

Comment:
This might be better in a function, to avoid the code size from bloating

Reply:
Good idea, will look into it

Reply:
Started to work on this, but it will bloat this already quite large PR further. I decided to do this in a follow up.